src/storage/txn/scheduler.rs

```diff
@@ -34,129 +34,123 @@
-        cid: u64,
-        cb_ctx: CbContext,
-        snapshot: EngineResult<E::Snap>,
-    },
     ReadFinished {
         cid: u64,
         pr: ProcessResult,
-    },
-    WritePrepareFinished {
-        cid: u64,
-        ctx: Context,
         tag: &'static str,
-        pr: ProcessResult,
-        to_be_write: Vec<Modify>,
-        rows: usize,
-    },
-    WritePrepareFailed {
-        cid: u64,
-        err: Error,
     },
     WriteFinished {
         cid: u64,
         pr: ProcessResult,
-        cb_ctx: CbContext,
         result: EngineResult<()>,
+        tag: &'static str,
+    },
+    FinishedWithErr {
+        cid: u64,
+        err: Error,
+        tag: &'static str,
     },
 }
 
 /// Debug for messages.
-impl<E: Engine> Debug for Msg<E> {
+impl Debug for Msg {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
         write!(f, "{}", self)
     }
 }
 
 /// Display for messages.
-impl<E: Engine> Display for Msg<E> {
+impl Display for Msg {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
         match *self {
             Msg::Quit => write!(f, "Quit"),
             Msg::RawCmd { ref cmd, .. } => write!(f, "RawCmd {:?}", cmd),
-            Msg::SnapshotFinished { cid, .. } => write!(f, "SnapshotFinished [cid={}]", cid),
             Msg::ReadFinished { cid, .. } => write!(f, "ReadFinished [cid={}]", cid),
-            Msg::WritePrepareFinished {
-                cid, ref ctx, tag, ..
-            } => write!(
-                f,
-                "WritePrepareFinished [cid={}, tag={}, ctx={:?}]",
-                cid, tag, ctx
-            ),
-            Msg::WritePrepareFailed { cid, ref err } => {
-                write!(f, "WritePrepareFailed [cid={}, err={:?}]", cid, err)
-            }
             Msg::WriteFinished { cid, .. } => write!(f, "WriteFinished [cid={}]", cid),
+            Msg::FinishedWithErr { cid, .. } => write!(f, "FinishedWithErr [cid={}]", cid),
         }
     }
 }
 
-/// Creates a callback to receive async results of write prepare from the storage engine.
-fn make_engine_cb<E: Engine>(
-    cmd: &'static str,
-    cid: u64,
-    pr: ProcessResult,
-    scheduler: worker::Scheduler<Msg<E>>,
-    rows: usize,
-) -> EngineCallback<()> {
-    Box::new(move |(cb_ctx, result)| {
-        match scheduler.schedule(Msg::WriteFinished {
-            cid,
-            pr,
-            cb_ctx,
-            result,
-        }) {
-            Ok(_) => {
-                KV_COMMAND_KEYWRITE_HISTOGRAM_VEC
-                    .with_label_values(&[cmd])
-                    .observe(rows as f64);
-            }
-            e @ Err(ScheduleError::Stopped(_)) => info!("scheduler worker stopped, {:?}", e),
-            Err(e) => {
-                panic!(
-                    "schedule WriteFinished msg failed, cid={}, err:{:?}",
-                    cid, e
-                );
-            }
+// TODO: Moves SchedEntity to Task once we adopt futures in the Engine trait.
+struct SchedEntity {
```

Comment:
`Entity` doesn't seem to be a good name, it makes more sense that task itself is an entity as it contains the actual task data. How about naming it as `TaskTracker` just like `RequestTracker` in endpoint.rs?